examples/developer-tools/src/controller.ts

```diff
@@ -318,27 +338,117 @@ export class Controller {
+          ', ',
+        )}.`;
+        if (!firstSuccessfulBlock) firstSuccessfulBlock = successes[0];
+      } else {
+        // Some success and some failure. Load the first success,
+        // and display the names of blocks that succeeded and failed to load.
+        messageDiv.textContent = `${file}: Successully loaded blocks named ${successes.join(
+          ', ',
+        )}, but failed to parse ${fails} blocks.`;
+        messageDiv.classList.add('warning-message');
+        if (!firstSuccessfulBlock) firstSuccessfulBlock = successes[0];
+      }
+
+      messages.appendChild(messageDiv);
+    }
+    if (firstSuccessfulBlock) {
+      loadBlock(this.mainWorkspace, firstSuccessfulBlock);
+    } else {
+      // No blocks were successfully parsed at all, so keep the file upload modal open
+      this.viewModel.toggleFileUploadWarning(true);
+      return;
+    }
+
+    this.viewModel.toggleFileUploadModal(false);
+    this.viewModel.toggleUploadResultsModal(true, messages);
   }
 
   /**
    * Given a file, tries to run it through our backwards-compatibility converter
-   * and load the block into block factory. If there's an error at any point,
-   * we show a warning and allow the user to try the upload again.
+   * and save the blocks in block factory.
    *
-   * @param file File containing the block json to load. This should be the file
-   *    downloaded directly from the old block factory tool.
+   * @param file File containing an array of block json to load. This should
+   * be the file downloaded directly from the old block factory tool.
+   * @returns a promise resolving to an object containing an array of successfully
+   * converted block names and the number of blocks that failed to load.
    */
-  private loadFromFile(file: File) {
-    file
-      .text()
-      .then((contents) => {
-        const fixedBlockData = convertBaseBlock(JSON.parse(contents));
-        loadBlockFromData(this.mainWorkspace, fixedBlockData);
-        this.viewModel.toggleFileUploadModal(false);
-      })
-      .catch((e) => {
-        this.viewModel.toggleFileUploadWarning(true);
-      });
+  private loadFromFile(
+    file: File,
+  ): Promise<{successes: string[]; fails: number}> {
+    const successes: string[] = [];
+    let fails = 0;
+    return file.text().then((contents) => {
+      try {
+        const blocksData = JSON.parse(contents);
+        if (!Array.isArray(blocksData) || blocksData.length === 0) {
+          // File did not contain a JSON array or was empty.
+          fails++;
+          return {successes, fails};
+        }
+        for (const block of blocksData) {
+          const fixedBlockData = convertBaseBlock(block);
+
+          // The name from the old factory might be used in this tool.
+          // Get a fresh name so we don't overwrite anything.
+          const name = storage.getNewUnusedName(fixedBlockData.fields.NAME);
+          fixedBlockData.fields.NAME = name;
```

Comment:
Possibly a bit fiddly to plumb back through to the results dialog, but might it be worth warning the user if the block(s) they are loading have been renamed?